File: frontend/lib/model/reservation.dart
class Reservation {
  final int id;
  final int utilisateurId;
  final int restaurantId;
  final DateTime dateReservation;
  final String heure;
  final int nombreCouverts;
  final String commentaire;
  final String statut;
  final String utilisateurNom;
  final String restaurantNom;

  Reservation({
    required this.id,
    required this.utilisateurId,
    required this.restaurantId,
    required this.dateReservation,
    required this.heure,
    required this.nombreCouverts,
    required this.commentaire,
    required this.statut,
    required this.utilisateurNom,
    required this.restaurantNom,
  });

  /// Convertit un JSON en objet Reservation
  factory Reservation.fromJson(Map<String, dynamic> json) {
    return Reservation(
      id: json['id'],
      utilisateurId: json['utilisateur_id'],
      restaurantId: json['restaurant_id'],
      dateReservation: DateTime.parse(json['date_reservation']),
      heure: json['heure'],
      nombreCouverts: json['nombre_couverts'],
      commentaire: json['commentaire'],
      statut: json['statut'],
      utilisateurNom: json['utilisateur_nom'],
      restaurantNom: json['restaurant_nom'],
    );
  }

  /// Convertit l'objet Reservation en JSON
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'utilisateur_id': utilisateurId,
      'restaurant_id': restaurantId,
      'date_reservation': dateReservation.toIso8601String(),
      'heure': heure,
      'nombre_couverts': nombreCouverts,
      'commentaire': commentaire,
      'statut': statut,
      'utilisateur_nom': utilisateurNom,
      'restaurant_nom': restaurantNom,
    };
  }
}

File: backend/models/reservation.js
const pool = require('../db');

class Reservation {
    // Vérifie si le restaurant est ouvert pour ce jour et cette heure
    static async checkAvailability(restaurant_id, date_reservation, heure) {
        try {
            console.log('🕒 Vérification disponibilité pour :', {
                restaurant_id,
                date_reservation,
                heure
            });

            const query = `
                SELECT *
                FROM ouverture
                WHERE restaurant_id = $1
                  AND TRIM(LOWER(jour)) = TRIM(LOWER(TO_CHAR($2::date, 'FMDay')))
                  AND heure_ouverture <= $3
                  AND heure_fermeture >= $3
            `;

            const { rows } = await pool.query(query, [
                restaurant_id,
                date_reservation,
                heure
            ]);

            console.log('📅 Créneaux trouvés :', rows.length);
            if (rows.length > 0) {
                console.log('✅ Le restaurant est ouvert.');
                return true;
            } else {
                console.warn('⚠️ Aucun créneau trouvé pour ce jour ou cette heure.');
                return false;
            }
        } catch (error) {
            console.error('❌ Erreur lors de la vérification de disponibilité :', error);
            throw error;
        }
    }

    // Création de réservation
    static async create({ utilisateur_id, restaurant_id, date_reservation, heure, nombre_couverts, commentaire }) {
    const isAvailable = await this.checkAvailability(restaurant_id, date_reservation, heure);

    if (!isAvailable) {
        throw new Error('Le restaurant est fermé à cette date ou heure. Choisissez un autre créneau.');
    }

    const query = `
        INSERT INTO reservations (utilisateur_id, restaurant_id, date_reservation, heure, nombre_couverts, commentaire)
        VALUES ($1, $2, $3, $4, $5, $6)
        RETURNING *
    `;

    const values = [utilisateur_id, restaurant_id, date_reservation, heure, nombre_couverts, commentaire];
    const { rows } = await pool.query(query, values);

    console.log('✅ Réservation créée avec succès :', rows[0]);
    return rows[0];
}

    // Récupère toutes les réservations
    static async findAll() {
        const { rows } = await pool.query(`
            SELECT r.*, u.nom AS utilisateur_nom, res.nom AS restaurant_nom
            FROM reservations r
            JOIN utilisateurs u ON r.utilisateur_id = u.id
            JOIN restaurant res ON r.restaurant_id = res.id
            ORDER BY date_reservation DESC, heure DESC
        `);
        return rows;
    }
    static async findByRestaurant(restaurant_id) {
    const { rows } = await pool.query(`
        SELECT r.*, u.nom AS utilisateur_nom, res.nom AS restaurant_nom
        FROM reservations r
        JOIN utilisateurs u ON r.utilisateur_id = u.id
        JOIN restaurant res ON r.restaurant_id = res.id
        WHERE r.restaurant_id = $1
        ORDER BY date_reservation DESC, heure DESC
    `, [restaurant_id]);
    return rows;
}

// Récupère toutes les réservations d'un utilisateur
static async findByUtilisateur(utilisateur_id) {
    const { rows } = await pool.query(`
        SELECT r.*, u.nom AS utilisateur_nom, res.nom AS restaurant_nom
        FROM reservations r
        JOIN utilisateurs u ON r.utilisateur_id = u.id
        JOIN restaurant res ON r.restaurant_id = res.id
        WHERE r.utilisateur_id = $1
        ORDER BY date_reservation DESC, heure DESC
    `, [utilisateur_id]);
    return rows;
}

static async valider(id) {
    try {
        const query = `
            UPDATE reservations
            SET statut = 'validée'
            WHERE id = $1
            RETURNING *;
        `;
        const { rows } = await pool.query(query, [id]);
        if (rows.length === 0) {
            throw new Error("Réservation introuvable.");
        }
        return rows[0];
    } catch (error) {
        console.error("❌ Erreur validation réservation :", error);
        throw error;
    }
}

// Dans class Reservation
static async refuser(id) {
    try {
        const query = `
            UPDATE reservations
            SET statut = 'refusée'
            WHERE id = $1
            RETURNING *;
        `;
        const { rows } = await pool.query(query, [id]);
        if (rows.length === 0) {
            throw new Error("Réservation introuvable.");
        }
        return rows[0];
    } catch (error) {
        console.error("❌ Erreur refus réservation :", error);
        throw error;
    }
}

}

module.exports = Reservation;
